gateway/src/main/kotlin/connection/DefaultGatewayConnection.kt

```diff
@@ -0,0 +1,282 @@
+package dev.kord.gateway.connection
+
+import dev.kord.gateway.*
+import dev.kord.gateway.connection.GatewayConnection.*
+import io.ktor.client.plugins.websocket.*
+import io.ktor.client.request.*
+import io.ktor.http.*
+import io.ktor.websocket.*
+import kotlinx.atomicfu.AtomicBoolean
+import kotlinx.atomicfu.AtomicRef
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.loop
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.MutableStateFlow
+import mu.KLogger
+import mu.KotlinLogging
+import java.io.ByteArrayOutputStream
+import java.io.Closeable
+import java.nio.charset.StandardCharsets
+import java.util.zip.InflaterOutputStream
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+import kotlin.time.TimeSource
+import kotlin.time.TimeSource.Monotonic.ValueTimeMark
+import io.ktor.websocket.CloseReason as KtorCloseReason
+
+/**
+ * Default implementation of [GatewayConnection].
+ */
+internal class DefaultGatewayConnection : GatewayConnection {
+
+    override val ping: MutableStateFlow<Duration?> = MutableStateFlow(null)
+    private lateinit var inflater: FrameInflater
+    private lateinit var data: Data
+    private lateinit var session: DefaultClientWebSocketSession
+    private lateinit var scope: CoroutineScope
+    private val atomicPossiblyZombie: AtomicBoolean = atomic(false)
+    private var possiblyZombie: Boolean by atomicPossiblyZombie
+    private val atomicSequence: AtomicRef<Int?> = atomic(null)
+    private val sequence: Int? by atomicSequence
+    private val log: KLogger = KotlinLogging.logger { }
+    private val atomicHeartbeatTimeMark: AtomicRef<ValueTimeMark> = atomic(TimeSource.Monotonic.markNow())
+    private var heartbeatTimeMark: ValueTimeMark by atomicHeartbeatTimeMark
+    private val atomicReceivedHello: AtomicBoolean = atomic(false)
+    private val atomicReadyData: AtomicRef<ReadyData?> = atomic(null)
+    private val atomicReconnectRequested: AtomicBoolean = atomic(false)
+    private val atomicInvalidSession: AtomicRef<InvalidSession?> = atomic(null)
+    private val atomicState: AtomicRef<State> = atomic(State.Uninitialized)
+    private val atomicManualClose: AtomicBoolean = atomic(false)
+    private val atomicHeartbeatJob: AtomicRef<Job?> = atomic(null)
+    private val sessionToken: String
+        get() = when (val sessionData = data.session) {
+            is Session.New -> sessionData.identify.token
+            is Session.Resumed -> sessionData.resume.token
+        }
+
+    override suspend fun open(data: Data): CloseReason {
+        if (!atomicState.compareAndSet(State.Uninitialized, State.Opening)) errorInvalidState()
+        this.data = data
+        val url = Url(data.uri)
+        val isZLibCompressed = url.parameters.contains("compress", "zlib-stream")
+        inflater = if (isZLibCompressed) FrameInflater.ZLib() else FrameInflater.None
+
+        coroutineScope {
```

Comment:
> what's the point of this coroutineScope here?

Look at 133, 163, 170. We are launching some jobs in the scope, and then, if the connection is closed, it will await until the jobs are done, before returning the close result